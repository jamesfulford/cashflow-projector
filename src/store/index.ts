import { signal } from "@preact/signals-react";

export const MINIMUM_DAYS = 365;
export const durationDaysState = signal(MINIMUM_DAYS);
