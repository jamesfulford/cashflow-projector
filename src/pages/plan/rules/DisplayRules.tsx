import ListGroup from "react-bootstrap/esm/ListGroup";
import sortBy from "lodash/sortBy";
import { EnhancedRule } from "./enhancedRules";
import { RulesDisplayProps } from "./RulesDisplay";
import { RuleDisplay } from "./RuleDisplay";

interface DisplayRulesProps extends RulesDisplayProps {
  rules: EnhancedRule[];
}
export function DisplayRules(props: DisplayRulesProps) {
  return (
    <div
      style={{
        overflowY: "scroll",
        height: "60vh",
        // TODO: make the entire left side scroll, with some stickiness; don't do scrolling here because 60vh is hacky
      }}
    >
      <ListGroup>
        {sortBy(props.rules, (r) => Math.abs(r.impact))
          .reverse()
          .map((rule, index) => {
            return (
              <RuleDisplay key={rule.id} rule={rule} index={index} {...props} />
            );
          })}
      </ListGroup>
    </div>
  );
}
