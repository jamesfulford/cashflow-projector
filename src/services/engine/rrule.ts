import { parseISO } from "date-fns/parseISO";
import { RRuleSet, rrulestr } from "rrule";
import { IApiRule } from "../../store/rules";

export function fromStringToDate(datestring: string): Date {
  return parseISO(datestring + "T00:00:00");
}
export function fromDateToString(date: Date): string {
  return date.toISOString().split("T")[0];
}

export function getDatesOfRRule(
  rrulestring: string,
  startDate: string,
  endDate: string,
): string[] {
  const rrule = rrulestr(rrulestring, { forceset: true }) as RRuleSet;

  // bug: `between` uses current time-of-day on all dates,
  // causing all the ex-dates to be ignored.
  // I've tried to tell it to use the right time of day, but with no luck.
  // so, here we are, implementing exdates manually.
  const exdates = new Set(rrule.exdates().map(fromDateToString));
  const dates = rrule
    .between(fromStringToDate(startDate), fromStringToDate(endDate), true)
    .map(fromDateToString)
    .filter((d) => !exdates.has(d));

  return dates;
}

export function getDatesOfRule(
  rule: IApiRule,
  startDate: string,
  endDate: string,
): string[] {
  return [
    ...getDatesOfRRule(rule.rrule, startDate, endDate),
    ...rule.exceptionalTransactions.map((t) => t.day),
  ].sort();
}
