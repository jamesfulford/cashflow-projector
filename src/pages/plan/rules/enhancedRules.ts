import { IApiRule, rulesState } from "../../../store/rules";
import {
  expenseSharesState,
  impactScoresState,
  rawImpactState,
} from "../../../store/ratios";
import { ReadonlySignal, computed } from "@preact/signals-core";
import { lastPaymentDayResultByRuleIDState } from "../../../store/computationDates";
import { computeLastPaymentDay } from "../../../services/engine/computeLastPaymentDate";

export type EnhancedRule = IApiRule & {
  impact: number;
  shareOfIncome: number;
  lastPaymentDayResult?: ReturnType<typeof computeLastPaymentDay>;
} & (
    | { isIncome: true; isExpense: false }
    | { isIncome: false; isExpense: true; shareOfExpenses: number }
  );
export const enhancedRulesState: ReadonlySignal<EnhancedRule[]> = computed(
  () => {
    const rules = rulesState.value;
    const sharesOfIncome = impactScoresState.value;
    const impacts = rawImpactState.value;
    const sharesOfExpenses = expenseSharesState.value;
    const lastPaymentDayResultByRuleID =
      lastPaymentDayResultByRuleIDState.value;

    return rules.map((r) => {
      const impact = impacts.get(r.id) ?? 0;
      const shareOfIncome = sharesOfIncome.get(r.id) ?? 0;
      const lastPaymentDayResult = lastPaymentDayResultByRuleID.get(r.id);

      const isExpense = impact <= 0;
      if (isExpense) {
        const shareOfExpenses = sharesOfExpenses.get(r.id) ?? 0;
        return {
          ...r,
          impact,
          shareOfIncome,
          isExpense: true,
          isIncome: false,
          shareOfExpenses,
          lastPaymentDayResult,
        };
      }
      return {
        ...r,
        impact,
        shareOfIncome,
        isExpense: false,
        isIncome: true,
        lastPaymentDayResult,
      };
    });
  },
);
