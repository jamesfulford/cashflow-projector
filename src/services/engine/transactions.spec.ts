import { test, expect } from "vitest";
import { computeTransactions } from "./transactions";
import { RuleType } from "../../store/rules";
import { RRule, RRuleSet } from "rrule";
import { fromStringToDate } from "./rrule";

test("should compute weekly rule transactions", () => {
  const transactions = computeTransactions(
    [
      {
        id: "rule1",
        name: "rule 1",
        type: RuleType.EXPENSE,
        version: 1,

        value: -10,
        rrule: new RRule({
          freq: RRule.WEEKLY,
          byweekday: [RRule.MO],
        }).toString(),
        exceptionalTransactions: [],
      },
    ],
    {
      startDate: "2024-05-19",
      endDate: "2024-06-19",
      currentBalance: 1000,
      setAside: 1000,
    },
  );
  expect(transactions).toMatchSnapshot();
});

test("should compute monthly rule transactions", () => {
  const transactions = computeTransactions(
    [
      {
        id: "rule1",
        name: "rule 1",
        type: RuleType.EXPENSE,
        version: 1,

        value: -10,
        rrule: new RRule({
          freq: RRule.MONTHLY,
          bymonthday: 1,
        }).toString(),
        exceptionalTransactions: [],
      },
    ],
    {
      startDate: "2024-05-19",
      endDate: "2024-09-19",
      currentBalance: 1000,
      setAside: 1000,
    },
  );
  expect(transactions).toMatchSnapshot();
});

test("should compute yearly rule transactions", () => {
  const transactions = computeTransactions(
    [
      {
        id: "rule1",
        name: "rule 1",
        type: RuleType.EXPENSE,
        version: 1,

        value: -10,
        rrule: new RRule({
          freq: RRule.YEARLY,
          dtstart: fromStringToDate("2024-06-01"),
        }).toString(),
        exceptionalTransactions: [],
      },
    ],
    {
      startDate: "2024-05-19",
      endDate: "2029-09-19",
      currentBalance: 1000,
      setAside: 1000,
    },
  );
  expect(transactions).toMatchSnapshot();
});

test("should honor count end type", () => {
  const transactions = computeTransactions(
    [
      {
        id: "rule1",
        name: "rule 1",
        type: RuleType.EXPENSE,
        version: 1,

        value: -10,
        rrule: new RRule({
          freq: RRule.WEEKLY,
          byweekday: [RRule.MO],
          count: 2,
        }).toString(),
        exceptionalTransactions: [],
      },
    ],
    {
      startDate: "2024-05-19",
      endDate: "2024-06-19",
      currentBalance: 1000,
      setAside: 1000,
    },
  );
  expect(transactions).toHaveLength(2);
  expect(transactions).toMatchSnapshot();
});

test("should honor until end type", () => {
  const transactions = computeTransactions(
    [
      {
        id: "rule1",
        name: "rule 1",
        type: RuleType.EXPENSE,
        version: 1,

        value: -10,
        rrule: new RRule({
          freq: RRule.WEEKLY,
          byweekday: [RRule.MO],
          until: fromStringToDate("2024-05-30"),
        }).toString(),
        exceptionalTransactions: [],
      },
    ],
    {
      startDate: "2024-05-19",
      endDate: "2024-06-19",
      currentBalance: 1000,
      setAside: 1000,
    },
  );
  expect(transactions[transactions.length - 1].day).not.toContain("2024-06");
  expect(transactions).toMatchSnapshot();
});

test("should honor exdates", () => {
  const rruleset = new RRuleSet();
  rruleset.rrule(
    new RRule({
      freq: RRule.WEEKLY,
      byweekday: [RRule.MO],
      until: fromStringToDate("2024-05-30"),
    }),
  );
  //   this is like skipping a transaction
  rruleset.exdate(fromStringToDate("2024-05-27"));

  const transactions = computeTransactions(
    [
      {
        id: "rule1",
        name: "rule 1",
        type: RuleType.EXPENSE,
        version: 1,

        value: -10,
        rrule: rruleset.toString(),
        exceptionalTransactions: [],
      },
    ],
    {
      startDate: "2024-05-19",
      endDate: "2024-05-30",
      currentBalance: 1000,
      setAside: 1000,
    },
  );
  expect(transactions[transactions.length - 1].day).not.toBe("2024-05-27");
  expect(transactions).toMatchSnapshot();
});

test("should honor exdates except for exceptionalTransactions", () => {
  const rruleset = new RRuleSet();
  rruleset.rrule(
    new RRule({
      freq: RRule.WEEKLY,
      byweekday: [RRule.MO],
      until: fromStringToDate("2024-05-30"),
    }),
  );
  rruleset.exdate(fromStringToDate("2024-05-27"));

  const transactions = computeTransactions(
    [
      {
        id: "rule1",
        name: "rule 1",
        type: RuleType.EXPENSE,
        version: 1,

        value: -10,
        rrule: rruleset.toString(),
        // having the exdate is similar to renaming/revaluing a transaction
        exceptionalTransactions: [
          {
            day: "2024-05-27",
            value: -20,
            name: "rule 1 exception",
            id: "exceptionalTransaction1",
          },
        ],
      },
    ],
    {
      startDate: "2024-05-19",
      endDate: "2024-05-30",
      currentBalance: 1000,
      setAside: 1000,
    },
  );
  expect(transactions[transactions.length - 1].day).toBe("2024-05-27");
  expect(transactions[transactions.length - 1].value).toBe(-20);
  expect(transactions[transactions.length - 1].name).toBe("rule 1 exception");
  expect(transactions[transactions.length - 1].exceptionalTransactionID).toBe(
    "exceptionalTransaction1",
  );
  expect(transactions).toMatchSnapshot();
});

test("should compute working_capital properly", () => {
  const transactions = computeTransactions(
    [
      {
        id: "rule1",
        name: "rule 1",
        type: RuleType.EXPENSE,
        version: 1,

        value: -10,
        rrule: new RRule({
          freq: RRule.WEEKLY,
          byweekday: [RRule.MO],
        }).toString(),
        exceptionalTransactions: [],
      },
      {
        id: "rule2",
        name: "rule 2",
        type: RuleType.INCOME,
        version: 1,

        value: 200,
        rrule: new RRule({
          freq: RRule.WEEKLY,
          interval: 2,
          dtstart: fromStringToDate("2024-05-01"),
          byweekday: [RRule.MO],
        }).toString(),
        exceptionalTransactions: [],
      },
    ],
    {
      startDate: "2024-05-19",
      endDate: "2024-08-19",
      currentBalance: 1000,
      setAside: 1000,
    },
  );
  expect(transactions).toMatchSnapshot();
});

test("should compute savings goal transactions", () => {
  const transactions = computeTransactions(
    [
      {
        id: "rule1",
        name: "rule 1",
        type: RuleType.SAVINGS_GOAL,
        version: 1,

        value: -10,
        rrule: new RRule({
          freq: RRule.WEEKLY,
          byweekday: [RRule.MO],
        }).toString(),
        exceptionalTransactions: [],

        progress: 902,
        goal: 1000,
      },
    ],
    {
      startDate: "2024-05-19",
      endDate: "2025-05-19",
      currentBalance: 1000,
      setAside: 1000,
    },
  );
  expect(transactions).toHaveLength(10);
  expect(transactions[transactions.length - 1].value).toBe(-8); // only what's needed to finish the goal
  expect(transactions).toMatchSnapshot();
});

test("should compute savings goal transactions even when goal is hit exactly", () => {
  const transactions = computeTransactions(
    [
      {
        id: "rule1",
        name: "rule 1",
        type: RuleType.SAVINGS_GOAL,
        version: 1,

        value: -10,
        rrule: new RRule({
          freq: RRule.WEEKLY,
          byweekday: [RRule.MO],
        }).toString(),
        exceptionalTransactions: [],

        progress: 980,
        goal: 1000,
      },
    ],
    {
      startDate: "2024-05-19",
      endDate: "2025-05-19",
      currentBalance: 1000,
      setAside: 1000,
    },
  );
  expect(transactions).toHaveLength(2);
  expect(transactions[transactions.length - 1].value).toBe(-10);
  expect(transactions[transactions.length - 1].isLastPayment).toBe(true);
  expect(transactions).toMatchSnapshot();
});

test("should compute savings goal transactions with exceptional transactions", () => {
  const transactions = computeTransactions(
    [
      {
        id: "rule1",
        name: "rule 1",
        type: RuleType.SAVINGS_GOAL,
        version: 1,

        value: -10,
        rrule: new RRule({
          freq: RRule.WEEKLY,
          byweekday: [RRule.MO],
        }).toString(),
        exceptionalTransactions: [
          { id: "exceptionalTransaction1", day: "2024-06-01", value: -50 },
        ],

        progress: 902,
        goal: 1000,
      },
    ],
    {
      startDate: "2024-05-19",
      endDate: "2025-05-19",
      currentBalance: 1000,
      setAside: 1000,
    },
  );
  expect(transactions).toHaveLength(6);
  expect(transactions).toMatchSnapshot();
});

test("should exclude exceptional transactions from savings goal if not needed", () => {
  const transactions = computeTransactions(
    [
      {
        id: "rule1",
        name: "rule 1",
        type: RuleType.SAVINGS_GOAL,
        version: 1,

        value: -10,
        rrule: new RRule({
          freq: RRule.WEEKLY,
          byweekday: [RRule.MO],
        }).toString(),
        exceptionalTransactions: [
          { id: "exceptionalTransaction1", day: "2024-08-01", value: -50 },
        ],

        progress: 902,
        goal: 1000,
      },
    ],
    {
      startDate: "2024-05-19",
      endDate: "2025-05-19",
      currentBalance: 1000,
      setAside: 1000,
    },
  );
  expect(transactions).toHaveLength(10);
  expect(transactions.map((t) => t.day)).not.toContain("2024-08-01");
  expect(transactions).toMatchSnapshot();
});

test("should include no transactions from savings goal if already met", () => {
  const transactions = computeTransactions(
    [
      {
        id: "rule1",
        name: "rule 1",
        type: RuleType.SAVINGS_GOAL,
        version: 1,

        value: -10,
        rrule: new RRule({
          freq: RRule.WEEKLY,
          byweekday: [RRule.MO],
        }).toString(),
        exceptionalTransactions: [],

        progress: 1000,
        goal: 1000,
      },
    ],
    {
      startDate: "2024-05-19",
      endDate: "2025-05-19",
      currentBalance: 1000,
      setAside: 1000,
    },
  );
  expect(transactions).toHaveLength(0);
});

test("should compute loan transactions", () => {
  const transactions = computeTransactions(
    [
      {
        id: "rule1",
        name: "rule 1",
        type: RuleType.LOAN,
        version: 1,

        value: -400,
        rrule: new RRule({
          freq: RRule.WEEKLY,
          byweekday: [RRule.MO],
        }).toString(),
        exceptionalTransactions: [],

        balance: 1000,
        interestRate: 0.08,
        minimumPayment: 0,
      },
    ],
    {
      startDate: "2024-05-19",
      endDate: "2034-05-19",
      currentBalance: 20000,
      setAside: 1000,
    },
  );
  expect(transactions).toHaveLength(3);
  expect(transactions[transactions.length - 1].value.toFixed(2)).toBe(
    "-201.39",
  ); // only what's needed to finish the goal; not -400
  expect(transactions).toMatchSnapshot();
});
