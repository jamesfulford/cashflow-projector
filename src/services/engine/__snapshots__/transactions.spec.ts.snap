// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`should compute monthly rule transactions 1`] = `
[
  {
    "calculations": {
      "balance": 990,
      "working_capital": -40,
    },
    "day": "2024-06-01",
    "id": "rule1::2024-06-01",
    "name": "rule 1",
    "rule_id": "rule1",
    "value": -10,
  },
  {
    "calculations": {
      "balance": 980,
      "working_capital": -40,
    },
    "day": "2024-07-01",
    "id": "rule1::2024-07-01",
    "name": "rule 1",
    "rule_id": "rule1",
    "value": -10,
  },
  {
    "calculations": {
      "balance": 970,
      "working_capital": -40,
    },
    "day": "2024-08-01",
    "id": "rule1::2024-08-01",
    "name": "rule 1",
    "rule_id": "rule1",
    "value": -10,
  },
  {
    "calculations": {
      "balance": 960,
      "working_capital": -40,
    },
    "day": "2024-09-01",
    "id": "rule1::2024-09-01",
    "name": "rule 1",
    "rule_id": "rule1",
    "value": -10,
  },
]
`;

exports[`should compute savings goal transactions 1`] = `
[
  {
    "calculations": {
      "balance": 990,
      "working_capital": -98,
    },
    "day": "2024-05-20",
    "id": "rule1::2024-05-20",
    "name": "rule 1",
    "rule_id": "rule1",
    "value": -10,
  },
  {
    "calculations": {
      "balance": 980,
      "working_capital": -98,
    },
    "day": "2024-05-27",
    "id": "rule1::2024-05-27",
    "name": "rule 1",
    "rule_id": "rule1",
    "value": -10,
  },
  {
    "calculations": {
      "balance": 970,
      "working_capital": -98,
    },
    "day": "2024-06-03",
    "id": "rule1::2024-06-03",
    "name": "rule 1",
    "rule_id": "rule1",
    "value": -10,
  },
  {
    "calculations": {
      "balance": 960,
      "working_capital": -98,
    },
    "day": "2024-06-10",
    "id": "rule1::2024-06-10",
    "name": "rule 1",
    "rule_id": "rule1",
    "value": -10,
  },
  {
    "calculations": {
      "balance": 950,
      "working_capital": -98,
    },
    "day": "2024-06-17",
    "id": "rule1::2024-06-17",
    "name": "rule 1",
    "rule_id": "rule1",
    "value": -10,
  },
  {
    "calculations": {
      "balance": 940,
      "working_capital": -98,
    },
    "day": "2024-06-24",
    "id": "rule1::2024-06-24",
    "name": "rule 1",
    "rule_id": "rule1",
    "value": -10,
  },
  {
    "calculations": {
      "balance": 930,
      "working_capital": -98,
    },
    "day": "2024-07-01",
    "id": "rule1::2024-07-01",
    "name": "rule 1",
    "rule_id": "rule1",
    "value": -10,
  },
  {
    "calculations": {
      "balance": 920,
      "working_capital": -98,
    },
    "day": "2024-07-08",
    "id": "rule1::2024-07-08",
    "name": "rule 1",
    "rule_id": "rule1",
    "value": -10,
  },
  {
    "calculations": {
      "balance": 910,
      "working_capital": -98,
    },
    "day": "2024-07-15",
    "id": "rule1::2024-07-15",
    "name": "rule 1",
    "rule_id": "rule1",
    "value": -10,
  },
  {
    "calculations": {
      "balance": 902,
      "working_capital": -98,
    },
    "day": "2024-07-22",
    "id": "rule1::2024-07-22",
    "isLastPayment": true,
    "name": "rule 1",
    "rule_id": "rule1",
    "value": -8,
  },
]
`;

exports[`should compute savings goal transactions even when goal is hit exactly 1`] = `
[
  {
    "calculations": {
      "balance": 990,
      "working_capital": -20,
    },
    "day": "2024-05-20",
    "id": "rule1::2024-05-20",
    "name": "rule 1",
    "rule_id": "rule1",
    "value": -10,
  },
  {
    "calculations": {
      "balance": 980,
      "working_capital": -20,
    },
    "day": "2024-05-27",
    "id": "rule1::2024-05-27",
    "isLastPayment": true,
    "name": "rule 1",
    "rule_id": "rule1",
    "value": -10,
  },
]
`;

exports[`should compute savings goal transactions with exceptional transactions 1`] = `
[
  {
    "calculations": {
      "balance": 990,
      "working_capital": -98,
    },
    "day": "2024-05-20",
    "id": "rule1::2024-05-20",
    "name": "rule 1",
    "rule_id": "rule1",
    "value": -10,
  },
  {
    "calculations": {
      "balance": 980,
      "working_capital": -98,
    },
    "day": "2024-05-27",
    "id": "rule1::2024-05-27",
    "name": "rule 1",
    "rule_id": "rule1",
    "value": -10,
  },
  {
    "calculations": {
      "balance": 930,
      "working_capital": -98,
    },
    "day": "2024-06-01",
    "exceptionalTransactionID": "exceptionalTransaction1",
    "id": "rule1::2024-06-01::exceptionalTransaction1",
    "name": "rule 1",
    "rule_id": "rule1",
    "value": -50,
  },
  {
    "calculations": {
      "balance": 920,
      "working_capital": -98,
    },
    "day": "2024-06-03",
    "id": "rule1::2024-06-03",
    "name": "rule 1",
    "rule_id": "rule1",
    "value": -10,
  },
  {
    "calculations": {
      "balance": 910,
      "working_capital": -98,
    },
    "day": "2024-06-10",
    "id": "rule1::2024-06-10",
    "name": "rule 1",
    "rule_id": "rule1",
    "value": -10,
  },
  {
    "calculations": {
      "balance": 902,
      "working_capital": -98,
    },
    "day": "2024-06-17",
    "id": "rule1::2024-06-17",
    "isLastPayment": true,
    "name": "rule 1",
    "rule_id": "rule1",
    "value": -8,
  },
]
`;

exports[`should compute weekly rule transactions 1`] = `
[
  {
    "calculations": {
      "balance": 990,
      "working_capital": -50,
    },
    "day": "2024-05-20",
    "id": "rule1::2024-05-20",
    "name": "rule 1",
    "rule_id": "rule1",
    "value": -10,
  },
  {
    "calculations": {
      "balance": 980,
      "working_capital": -50,
    },
    "day": "2024-05-27",
    "id": "rule1::2024-05-27",
    "name": "rule 1",
    "rule_id": "rule1",
    "value": -10,
  },
  {
    "calculations": {
      "balance": 970,
      "working_capital": -50,
    },
    "day": "2024-06-03",
    "id": "rule1::2024-06-03",
    "name": "rule 1",
    "rule_id": "rule1",
    "value": -10,
  },
  {
    "calculations": {
      "balance": 960,
      "working_capital": -50,
    },
    "day": "2024-06-10",
    "id": "rule1::2024-06-10",
    "name": "rule 1",
    "rule_id": "rule1",
    "value": -10,
  },
  {
    "calculations": {
      "balance": 950,
      "working_capital": -50,
    },
    "day": "2024-06-17",
    "id": "rule1::2024-06-17",
    "name": "rule 1",
    "rule_id": "rule1",
    "value": -10,
  },
]
`;

exports[`should compute working_capital properly 1`] = `
[
  {
    "calculations": {
      "balance": 990,
      "working_capital": -20,
    },
    "day": "2024-05-20",
    "id": "rule1::2024-05-20",
    "name": "rule 1",
    "rule_id": "rule1",
    "value": -10,
  },
  {
    "calculations": {
      "balance": 980,
      "working_capital": -20,
    },
    "day": "2024-05-27",
    "id": "rule1::2024-05-27",
    "name": "rule 1",
    "rule_id": "rule1",
    "value": -10,
  },
  {
    "calculations": {
      "balance": 1180,
      "working_capital": 160,
    },
    "day": "2024-05-27",
    "id": "rule2::2024-05-27",
    "name": "rule 2",
    "rule_id": "rule2",
    "value": 200,
  },
  {
    "calculations": {
      "balance": 1170,
      "working_capital": 160,
    },
    "day": "2024-06-03",
    "id": "rule1::2024-06-03",
    "name": "rule 1",
    "rule_id": "rule1",
    "value": -10,
  },
  {
    "calculations": {
      "balance": 1160,
      "working_capital": 160,
    },
    "day": "2024-06-10",
    "id": "rule1::2024-06-10",
    "name": "rule 1",
    "rule_id": "rule1",
    "value": -10,
  },
  {
    "calculations": {
      "balance": 1360,
      "working_capital": 340,
    },
    "day": "2024-06-10",
    "id": "rule2::2024-06-10",
    "name": "rule 2",
    "rule_id": "rule2",
    "value": 200,
  },
  {
    "calculations": {
      "balance": 1350,
      "working_capital": 340,
    },
    "day": "2024-06-17",
    "id": "rule1::2024-06-17",
    "name": "rule 1",
    "rule_id": "rule1",
    "value": -10,
  },
  {
    "calculations": {
      "balance": 1340,
      "working_capital": 340,
    },
    "day": "2024-06-24",
    "id": "rule1::2024-06-24",
    "name": "rule 1",
    "rule_id": "rule1",
    "value": -10,
  },
  {
    "calculations": {
      "balance": 1540,
      "working_capital": 520,
    },
    "day": "2024-06-24",
    "id": "rule2::2024-06-24",
    "name": "rule 2",
    "rule_id": "rule2",
    "value": 200,
  },
  {
    "calculations": {
      "balance": 1530,
      "working_capital": 520,
    },
    "day": "2024-07-01",
    "id": "rule1::2024-07-01",
    "name": "rule 1",
    "rule_id": "rule1",
    "value": -10,
  },
  {
    "calculations": {
      "balance": 1520,
      "working_capital": 520,
    },
    "day": "2024-07-08",
    "id": "rule1::2024-07-08",
    "name": "rule 1",
    "rule_id": "rule1",
    "value": -10,
  },
  {
    "calculations": {
      "balance": 1720,
      "working_capital": 700,
    },
    "day": "2024-07-08",
    "id": "rule2::2024-07-08",
    "name": "rule 2",
    "rule_id": "rule2",
    "value": 200,
  },
  {
    "calculations": {
      "balance": 1710,
      "working_capital": 700,
    },
    "day": "2024-07-15",
    "id": "rule1::2024-07-15",
    "name": "rule 1",
    "rule_id": "rule1",
    "value": -10,
  },
  {
    "calculations": {
      "balance": 1700,
      "working_capital": 700,
    },
    "day": "2024-07-22",
    "id": "rule1::2024-07-22",
    "name": "rule 1",
    "rule_id": "rule1",
    "value": -10,
  },
  {
    "calculations": {
      "balance": 1900,
      "working_capital": 880,
    },
    "day": "2024-07-22",
    "id": "rule2::2024-07-22",
    "name": "rule 2",
    "rule_id": "rule2",
    "value": 200,
  },
  {
    "calculations": {
      "balance": 1890,
      "working_capital": 880,
    },
    "day": "2024-07-29",
    "id": "rule1::2024-07-29",
    "name": "rule 1",
    "rule_id": "rule1",
    "value": -10,
  },
  {
    "calculations": {
      "balance": 1880,
      "working_capital": 880,
    },
    "day": "2024-08-05",
    "id": "rule1::2024-08-05",
    "name": "rule 1",
    "rule_id": "rule1",
    "value": -10,
  },
  {
    "calculations": {
      "balance": 2080,
      "working_capital": 1070,
    },
    "day": "2024-08-05",
    "id": "rule2::2024-08-05",
    "name": "rule 2",
    "rule_id": "rule2",
    "value": 200,
  },
  {
    "calculations": {
      "balance": 2070,
      "working_capital": 1070,
    },
    "day": "2024-08-12",
    "id": "rule1::2024-08-12",
    "name": "rule 1",
    "rule_id": "rule1",
    "value": -10,
  },
  {
    "calculations": {
      "balance": 2270,
      "working_capital": 1270,
    },
    "day": "2024-08-19",
    "id": "rule2::2024-08-19",
    "name": "rule 2",
    "rule_id": "rule2",
    "value": 200,
  },
]
`;

exports[`should compute yearly rule transactions 1`] = `
[
  {
    "calculations": {
      "balance": 990,
      "working_capital": -60,
    },
    "day": "2024-06-01",
    "id": "rule1::2024-06-01",
    "name": "rule 1",
    "rule_id": "rule1",
    "value": -10,
  },
  {
    "calculations": {
      "balance": 980,
      "working_capital": -60,
    },
    "day": "2025-06-01",
    "id": "rule1::2025-06-01",
    "name": "rule 1",
    "rule_id": "rule1",
    "value": -10,
  },
  {
    "calculations": {
      "balance": 970,
      "working_capital": -60,
    },
    "day": "2026-06-01",
    "id": "rule1::2026-06-01",
    "name": "rule 1",
    "rule_id": "rule1",
    "value": -10,
  },
  {
    "calculations": {
      "balance": 960,
      "working_capital": -60,
    },
    "day": "2027-06-01",
    "id": "rule1::2027-06-01",
    "name": "rule 1",
    "rule_id": "rule1",
    "value": -10,
  },
  {
    "calculations": {
      "balance": 950,
      "working_capital": -60,
    },
    "day": "2028-06-01",
    "id": "rule1::2028-06-01",
    "name": "rule 1",
    "rule_id": "rule1",
    "value": -10,
  },
  {
    "calculations": {
      "balance": 940,
      "working_capital": -60,
    },
    "day": "2029-06-01",
    "id": "rule1::2029-06-01",
    "name": "rule 1",
    "rule_id": "rule1",
    "value": -10,
  },
]
`;

exports[`should exclude exceptional transactions from savings goal if not needed 1`] = `
[
  {
    "calculations": {
      "balance": 990,
      "working_capital": -98,
    },
    "day": "2024-05-20",
    "id": "rule1::2024-05-20",
    "name": "rule 1",
    "rule_id": "rule1",
    "value": -10,
  },
  {
    "calculations": {
      "balance": 980,
      "working_capital": -98,
    },
    "day": "2024-05-27",
    "id": "rule1::2024-05-27",
    "name": "rule 1",
    "rule_id": "rule1",
    "value": -10,
  },
  {
    "calculations": {
      "balance": 970,
      "working_capital": -98,
    },
    "day": "2024-06-03",
    "id": "rule1::2024-06-03",
    "name": "rule 1",
    "rule_id": "rule1",
    "value": -10,
  },
  {
    "calculations": {
      "balance": 960,
      "working_capital": -98,
    },
    "day": "2024-06-10",
    "id": "rule1::2024-06-10",
    "name": "rule 1",
    "rule_id": "rule1",
    "value": -10,
  },
  {
    "calculations": {
      "balance": 950,
      "working_capital": -98,
    },
    "day": "2024-06-17",
    "id": "rule1::2024-06-17",
    "name": "rule 1",
    "rule_id": "rule1",
    "value": -10,
  },
  {
    "calculations": {
      "balance": 940,
      "working_capital": -98,
    },
    "day": "2024-06-24",
    "id": "rule1::2024-06-24",
    "name": "rule 1",
    "rule_id": "rule1",
    "value": -10,
  },
  {
    "calculations": {
      "balance": 930,
      "working_capital": -98,
    },
    "day": "2024-07-01",
    "id": "rule1::2024-07-01",
    "name": "rule 1",
    "rule_id": "rule1",
    "value": -10,
  },
  {
    "calculations": {
      "balance": 920,
      "working_capital": -98,
    },
    "day": "2024-07-08",
    "id": "rule1::2024-07-08",
    "name": "rule 1",
    "rule_id": "rule1",
    "value": -10,
  },
  {
    "calculations": {
      "balance": 910,
      "working_capital": -98,
    },
    "day": "2024-07-15",
    "id": "rule1::2024-07-15",
    "name": "rule 1",
    "rule_id": "rule1",
    "value": -10,
  },
  {
    "calculations": {
      "balance": 902,
      "working_capital": -98,
    },
    "day": "2024-07-22",
    "id": "rule1::2024-07-22",
    "isLastPayment": true,
    "name": "rule 1",
    "rule_id": "rule1",
    "value": -8,
  },
]
`;

exports[`should honor count end type 1`] = `
[
  {
    "calculations": {
      "balance": 990,
      "working_capital": -20,
    },
    "day": "2024-05-20",
    "id": "rule1::2024-05-20",
    "name": "rule 1",
    "rule_id": "rule1",
    "value": -10,
  },
  {
    "calculations": {
      "balance": 980,
      "working_capital": -20,
    },
    "day": "2024-05-27",
    "id": "rule1::2024-05-27",
    "name": "rule 1",
    "rule_id": "rule1",
    "value": -10,
  },
]
`;

exports[`should honor exdates 1`] = `
[
  {
    "calculations": {
      "balance": 990,
      "working_capital": -10,
    },
    "day": "2024-05-20",
    "id": "rule1::2024-05-20",
    "name": "rule 1",
    "rule_id": "rule1",
    "value": -10,
  },
]
`;

exports[`should honor exdates except for exceptionalTransactions 1`] = `
[
  {
    "calculations": {
      "balance": 990,
      "working_capital": -30,
    },
    "day": "2024-05-20",
    "id": "rule1::2024-05-20",
    "name": "rule 1",
    "rule_id": "rule1",
    "value": -10,
  },
  {
    "calculations": {
      "balance": 970,
      "working_capital": -30,
    },
    "day": "2024-05-27",
    "exceptionalTransactionID": "exceptionalTransaction1",
    "id": "rule1::2024-05-27::exceptionalTransaction1",
    "name": "rule 1 exception",
    "rule_id": "rule1",
    "value": -20,
  },
]
`;

exports[`should honor until end type 1`] = `
[
  {
    "calculations": {
      "balance": 990,
      "working_capital": -20,
    },
    "day": "2024-05-20",
    "id": "rule1::2024-05-20",
    "name": "rule 1",
    "rule_id": "rule1",
    "value": -10,
  },
  {
    "calculations": {
      "balance": 980,
      "working_capital": -20,
    },
    "day": "2024-05-27",
    "id": "rule1::2024-05-27",
    "name": "rule 1",
    "rule_id": "rule1",
    "value": -10,
  },
]
`;
